third_party/move/move-compiler-v2/src/env_pipeline/inlining_optimization.rs

```diff
@@ -0,0 +1,695 @@
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module implements the inlining optimization, whose aim is to reduce the number
+//! of function calls (which comes at the cost of increasing code size).
+//!
+//! See the documentation for the `optimize` function for more details on what inlining is.
+
+use crate::{
+    env_pipeline::rewrite_target::{RewriteState, RewriteTarget, RewriteTargets, RewritingScope},
+    file_format_generator::MAX_LOCAL_COUNT,
+};
+use codespan_reporting::diagnostic::Severity;
+use move_binary_format::file_format::Visibility;
+use move_model::{
+    ast::{Exp, ExpData, Operation, Pattern, TempIndex},
+    exp_rewriter::ExpRewriterFunctions,
+    metadata::LanguageVersion,
+    model::{
+        FunId, FunctionEnv, FunctionSize, GlobalEnv, Loc, ModuleEnv, ModuleId, NodeId, Parameter,
+        QualifiedId,
+    },
+    ty::Type,
+};
+use petgraph::{algo::kosaraju_scc, prelude::DiGraphMap};
+use std::collections::BTreeSet;
+
+/// [TODO]: tune the heuristic limits below
+/// A conservative heuristic limit posed by the inlining optimization on how
+/// large a caller function can grow to due to inlining.
+const MAX_CALLER_CODE_SIZE: usize = 1024;
+/// A conservative heuristic limit posed by the inlining optimization on how
+/// large a callee function can be for it to be considered for inlining.
+const MAX_CALLEE_CODE_SIZE: usize = 128;
+/// Number of times we want to apply "unrolling" of functions with inlining.
+const UNROLL_DEPTH: usize = 10;
+
+/// Optimize functions in target modules by applying inlining transformations.
+/// With inlining, a call site of the form:
+/// ```move
+/// foo(a, b, c, ...)
+/// ```
+/// with
+/// ```move
+/// fun foo(x, y, z, ...) { body }
+/// ```
+/// becomes:
+/// ```move
+/// let (x, y, z, ...) = (a, b, c, ...); body
+/// ```
+///
+/// The `across_package` value controls if inlining is performed across package boundaries.
+/// With across-package inlining, calls to functions in other packages may be inlined,
+/// which means that if the other package is upgraded, one would get the behavior
+/// at the inline-time rather than the latest upgrade.
+pub fn optimize(env: &mut GlobalEnv, across_package: bool) {
+    let mut targets = RewriteTargets::create(env, RewritingScope::CompilationTarget);
+    let skip_functions = find_cycles_in_call_graph(env, &targets);
+    targets.filter(|target, _| {
+        if let RewriteTarget::MoveFun(function_id) = target {
+            let function = env.get_function(*function_id);
+            // We will consider inlining the callees in a function on if it satisfies all of:
+            // - is not a part of a cycle in the call graph
+            // - is in a primary target module
+            // - is not in a script module
+            // - is not a test only function
+            // - is not a verify only function
+            // - is not a native function
+            // - is not an inline function
+            !skip_functions.contains(function_id)
```

Comment:
Since you incrementally unroll, could do those as well? Maybe later if certain config is enabled.

Reply:
Yes, I checked that limited unrolling of recursive functions works correctly (with the added example for recursive fibonacci in this PR), but it usually results in bloat with not much benefit. So disabled it by default (however we can easily enable it in the future).